sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Cast.java

```diff
@@ -0,0 +1,582 @@
+  }
+
+  @VisibleForTesting
+  static Row castRow(Row input, Schema inputSchema, Schema outputSchema) {
+    Row.Builder output = Row.withSchema(outputSchema);
+    for (int i = 0; i < outputSchema.getFieldCount(); i++) {
+      Schema.Field outputField = outputSchema.getField(i);
+
+      int fromFieldIdx = inputSchema.indexOf(outputField.getName());
+      Schema.Field inputField = inputSchema.getField(fromFieldIdx);
+
+      Object inputValue = input.getValue(fromFieldIdx);
+
+      if (inputValue == null) {
+        output.addValue(null);
+      } else {
+        Object value = castObject(inputValue, inputField.getType(), outputField.getType());
+        output.addValue(value);
+      }
+    }
+
+    return output.build();
+  }
+
+  @VisibleForTesting
+  @SuppressWarnings("unchecked")
+  static Object castObject(
+      Object inputValue, Schema.FieldType inputFieldType, Schema.FieldType outputFieldType) {
+
+    if (inputFieldType.getTypeName() == ROW) {
+      return castRow(
+          (Row) inputValue, inputFieldType.getRowSchema(), outputFieldType.getRowSchema());
+    } else if (inputFieldType.getTypeName() == ARRAY) {
+      List<Object> inputValues = (List<Object>) inputValue;
+      List<Object> outputValues = new ArrayList<>(inputValues.size());
+
+      for (Object elem : inputValues) {
+        outputValues.add(
+            castObject(
+                elem,
+                inputFieldType.getCollectionElementType(),
+                outputFieldType.getCollectionElementType()));
+      }
+
+      return outputValues;
+    } else if (inputFieldType.getTypeName() == MAP) {
+      Map<Object, Object> inputMap = (Map<Object, Object>) inputValue;
+      Map<Object, Object> outputMap = Maps.newHashMapWithExpectedSize(inputMap.size());
+
+      for (Map.Entry<Object, Object> entry : inputMap.entrySet()) {
+        Object outputKey =
+            castObject(
+                entry.getKey(), inputFieldType.getMapKeyType(), outputFieldType.getMapKeyType());
+        Object outputValue =
+            castObject(
+                entry.getValue(),
+                inputFieldType.getMapValueType(),
+                outputFieldType.getMapValueType());
+
+        outputMap.put(outputKey, outputValue);
+      }
+
+      return outputMap;
+    } else {
+      if (inputFieldType.getTypeName().equals(outputFieldType.getTypeName())) {
+        return inputValue;
+      } else {
+        KV<TypeName, TypeName> key =
+            KV.of(inputFieldType.getTypeName(), outputFieldType.getTypeName());
+        return Inference.ALL_MAP.get(key).apply(inputValue);
+      }
+    }
+  }
+
+  /** Configures casting for nullable fields. */
+  public enum Nullability {
+    /** Can cast only if nullability is the same. */
+    SAME,
+    /** Can cast non-nullable fields to nullable. */
+    WEAKEN,
+    /** Can cast nullable fields to non-nullable and forth. */
+    IGNORE
+  }
+
+  /** Configures casting of primitive types. */
+  public enum Type {
```

Comment:
Do you need all the knobs? I would imagine that if the user requests to cast from A to B then there is just one good interpretation. So all these enums would be not needed.

Reply:
There is a need, but I'm not sure it's organized properly. Motivation is to separate different kinds of conversions and provide validation not to accidentally run unsafe code that can crash in runtime.

Reply:
I believe that casting from type S to type T you can statically determine whether error handling is required, and what sort.

Reply:
Yes, that's was the intention, explicit opt-in for unsafe operations, otherwise, PTransform throws an exception during expand.